compiler/rustc_hir_analysis/src/check/check.rs

```diff
@@ -1115,20 +1114,38 @@ pub(super) fn check_transparent<'tcx>(tcx: TyCtxt<'tcx>, adt: ty::AdtDef<'tcx>)
             }
         }
 
-        (span, zst, align, check_non_exhaustive(tcx, ty).break_value())
+        (span, zst, layout, check_non_exhaustive(tcx, ty).break_value())
     });
 
-    let non_zst_fields = field_infos
+    let non_layout_fields = field_infos
         .clone()
-        .filter_map(|(span, zst, _align, _non_exhaustive)| if !zst { Some(span) } else { None });
+        .filter_map(|(span, _, layout, _)| if layout.is_err() { Some(span) } else { None });
+    let non_layout_count = non_layout_fields.clone().count();
+    let non_zst_fields =
+        field_infos.clone().filter_map(
+            |(span, zst, layout, _)| if layout.is_ok() && !zst { Some(span) } else { None },
+        );
```

Comment:
I'm debating whether or not extra comments are warranted in this computation.

You have an awesome explanation of the situation in your PR description, and I'm just wondering the code itself will be overly obtuse without some comments that reflect pieces of that description...

Reply:
I now added a short explanation on why we gather the data and which properties we have to check for.